clash-cores/src/Clash/Cores/Xilinx/VIO/Internal/BlackBoxes.hs

```diff
@@ -0,0 +1,292 @@
+          _         -> pure outProbes
+
+  | otherwise = error $ "vioProbeBBTF, bad bbCtx: " <> show bbCtx
+
+vioProbeTclTF :: HasCallStack => TemplateFunction
+vioProbeTclTF =
+  TemplateFunction
+    (
+--      0 :       -- KnownDomain dom
+--      1 :       -- VIO dom a o
+      2 :       -- initialOutputProbeValues
+      3 :       -- Clock dom
+      [4,5..]   -- input probes
+    )
+    -- 'validateVioProbeBCC' already produces string describing
+    -- failing checks, but 'TemplateFunction' cannot handle this
+    -- yet. This is prepared to get updated easily as soon as the
+    -- feature gets implemented in clash-lib.
+    (maybe True error . validateVioProbeBBC)
+    vioProbeTclBBTF
+
+vioProbeTclBBTF ::
+  forall s.
+  (HasCallStack, Backend s) =>
+  BlackBoxContext ->
+  State s Doc
+vioProbeTclBBTF bbCtx
+  | ( _knownDomainDom
+    : _vioConstraint
+    : initialOutputProbeValues
+    : _clk
+    : _inputProbes
+    ) <- map fst $ DSL.tInputs bbCtx
+  , [vioProbeName] <- bbQsysIncName bbCtx
+  , Right (inTys, outTys) <- probesFromTypes bbCtx
+  = do
+
+    let
+      outIVals = case DSL.ety initialOutputProbeValues of
+        CustomProduct{} ->
+          -- TODO: support custom bit representations for S&P types
+          error "Custom bit representations are not supported within VIOs yet"
+        _ ->
+          constantProbeValues
+            (DSL.ety initialOutputProbeValues)
+            (DSL.eex initialOutputProbeValues)
+
+    when (length outIVals /= length outTys) $
+      error "Internal: Unexpected Structure (this should not happen)"
+
+    pure $ renderTcl $ pure $ IpConfigPurpose
+      $ (defIpConfig "vio" "3.0" vioProbeName)
+           { properties = vioProbeProperties inTys outTys outIVals
+           }
+  | otherwise = error $ "vioProbeBBTF, bad bbCtx: " <> show bbCtx
+ where
+  vioProbeProperties inTys outTys outIVals =
+    [ property @Int "C_NUM_PROBE_IN" $ length inTys
+    , property @Int "C_NUM_PROBE_OUT" $ length outTys
+    ] <>
+    [ property @Int (T.concat ["C_PROBE_IN", T.pack $ show i, "_WIDTH"]) n
+    | (i,t) <- zip [0 :: Int, 1..] inTys
+    , let n = fromInteger $ DSL.tySize t
+    ] <>
+    [ property @Int (T.concat ["C_PROBE_OUT", T.pack $ show i, "_WIDTH"]) n
+    | (i,t) <- zip [0 :: Int, 1..] outTys
+    , let n = fromInteger $ DSL.tySize t
+    ] <>
+    [ property @Int (T.concat ["C_PROBE_OUT", T.pack $ show i, "_INIT_VAL"]) v
+    | (i, v) <- zip [0 :: Int, 1..] outIVals
+    ]
+
+-- | Turns a known to be constant value 'Expr' of some known 'HWType'
+-- into a list of initial probe values, where the probe distribution
+-- is determined according to the 'HWType'.
+constantProbeValues :: HasCallStack => HWType -> Expr -> [Int]
+constantProbeValues ty expr = case bits (DSL.tySize ty) expr of
+  Left e ->
+    error
+      $ "Cannot get constant initial output probe values " <>
+        "from the following expression:\n" <> show e
```

Comment:
Use `showPpr` and the `[I.i` quasiquoter for much nicer error messages, see:

https://github.com/clash-lang/clash-compiler/blob/031a664b87042969996307d3cabcabed94d81f55/clash-lib/src/Clash/Core/TermLiteral.hs#L164-L180

We should probably also point out that users can use `$(lift ....)` to make TH do the reduction instead of Clash.

Reply:
`Expr`'s are not pretty printable, but we can `show` the full and sub-expressions at least.